### PR TITLE
[spring] Errors

### PR DESCRIPTION
**Pull request by mh-park**
This adds a `details` field to `yarpcerror.Status` and renames it to `yarpcerror.encodingError`. (Note that it is now private). Consequently, `yarpcerror.encodingError` should be constructed via `yarpcerror.New(...)`.

We also pull out the `code`, `message`, and `name` to `yarpcerror.Info`. This allows the `Info` to be copied around and used in parallel with other error types.

---

### Current direction:
We want a uniform way the encoding handler returns a body or error (keyword or: we want to only return one). Currently, that diverged a little bit with thrift and protobuf; with the introduction of encoding middleware and the `EncodingToTransportProcedure(..)` method, we do something that is unidiomatic-go with thrift by returning a response *and* an error, in the case of an application error. Our solution that we decided on to fix this is:
#### Encoding
- Thrift: with the details field in `yarpcerror.status`, we move the response `wire.Value` into the error details field, in case of an application error. Everything error related now travels within a `yarpcerror.status`, and the response body is nil.
- Protobuf: `yarpcerror.status` is built around the `proto.Status` from gRPC. We expect this to be straightforward.
- JSON: ???
#### Transport
- TChannel: TChannel sends errors over the response body. To keep wire-compatibility, we now need to make sure to retrieve the application error from `yarpcerror.status` and send it over the response body. At the transport level, we don't actually know if this details field is a Thrift exception or Proto details, so we just always send the details field in the response body. (This means that for Protobuf, we will be sending the status details over the body.)
- gRPC: gRPC sends errors over headers. We will always send `yarpcerror.status` over gRPC as if it is a `proto.Status`. (This means using Thrift with gRPC will end up sending application errors over headers.)
- HTTP: ??!??!

#### TL;DR
Encoding handler is responsible for wrapping all application errors into a `yarpcerror.encodingError`, and only returning a body xor error.
Transport handler is responsible for handling the body or error returned from the encoding handler accordingly.

---

Feel free to edit this as we go! I just wanted to document thoughts somewhere before I forget where we're headed.

### CODE REVIEW

**Comment by mh-park** [comment]
Please send some opinions/ideas/thoughts!

**Comment by kriskowal** [comment]
This looks good.

For JSON, we can have JSON encoded error details and treat them the same way we would for error details in general at the transport level. I imagine this involves calling Procedure.EncodeError(details) just like it would for Thrift or Protobuf. I also imagine that gRPC will make a special case for the "protobuf" Procedure.Encoding.

For HTTP, we’ll have to have a conversation about the Accept-Both header and how it affects how errors are wired. It won’t be a problem. The body carries the error message by default. If the client and server agree, there’s a newer protocol that more closely resembles gRPC.

File: v2/internal/internalyarpcerror/yarpcerrors.go [hunk 1]
```diff
@@ -1,43 +0,0 @@
-// Copyright (c) 2018 Uber Technologies, Inc.
```
Comment (AlexAPB) [inline v2/internal/internalyarpcerror/yarpcerrors.go]: 💯

File: v2/yarpcerror/errors.go [hunk 2]
```diff
@@ -25,31 +25,52 @@ import (
...
-func Newf(code Code, format string, args ...interface{}) *Status {
+func Newf(code Code, format string, args ...interface{}) error {
 	if code == CodeOK {
 		return nil
 	}
-	return &Status{
+	return New(code, sprintf(format, args...))
+}
+
+// New return a new Status.
```
Comment (AlexAPB) [inline v2/yarpcerror/errors.go]: drop `Status` since it's private now

File: v2/yarpcerror/errors.go [hunk 3]
```diff
@@ -25,31 +25,52 @@ import (
...
+//
+// The Code should never be CodeOK; if it is, this will return nil.
```
Comment (AlexAPB) [inline v2/yarpcerror/errors.go]: Nit: `New will return nil if given 'CodeOK'.` 

Saying it should never be CodeOK seems unnecessary.

File: v2/yarpcerror/errors.go [hunk 4]
```diff
@@ -25,31 +25,52 @@ import (
...
+	status := &status{
 		code:    code,
-		message: sprintf(format, args...),
+		message: message,
 	}
+	for _, opt := range options {
+		opt.apply(status)
+	}
+
+	// The name must only contain lowercase letters from a-z and dashes (-), and
```
Comment (AlexAPB) [inline v2/yarpcerror/errors.go]: nit: can we move this documentation to the function declaration for `validateName`?

File: v2/yarpcerror/errors.go [hunk 5]
```diff
@@ -25,31 +25,52 @@ import (
 	"fmt"
 )
 
-// Newf returns a new Status.
+// Newf returns a new Status, with message formatting.
```
Comment (AlexAPB) [inline v2/yarpcerror/errors.go]: drop Status

File: v2/yarpcerror/errors.go [hunk 6]
```diff
@@ -59,71 +80,70 @@ func FromError(err error) *Status {
...
-		code:    s.code,
-		name:    name,
-		message: s.message,
-	}
+func WithName(name string) StatusOption {
+	return StatusOption{func(s *status) { s.name = name }}
+}
+
+// WithDetails adds semantic metadata to a Status.
+func WithDetails(details interface{}) StatusOption {
```
Comment (AlexAPB) [inline v2/yarpcerror/errors.go]: Thoughts on just calling this `WithApplicationError`?

Even though we borrowed this from gRPC, I'm not a huge fan of the term `Details`.

File: v2/yarpcerror/errors.go [hunk 7]
```diff
@@ -59,71 +80,70 @@ func FromError(err error) *Status {
...
 	name    string
 	message string
+	details interface{}
 }
 
-// WithName returns a new Status with the given name.
+// StatusOption provides options that may be called to the constructor.
+type StatusOption struct{ apply func(*status) }
+
+// WithName returns a new status with the given name.
```
Comment (AlexAPB) [inline v2/yarpcerror/errors.go]: drop status

File: v2/yarpcerror/errors.go [hunk 8]
```diff
@@ -59,71 +80,70 @@ func FromError(err error) *Status {
...
+type status struct {
 	code    Code
 	name    string
 	message string
+	details interface{}
 }
 
-// WithName returns a new Status with the given name.
+// StatusOption provides options that may be called to the constructor.
+type StatusOption struct{ apply func(*status) }
```
Comment (AlexAPB) [inline v2/yarpcerror/errors.go]: maybe just `ErrorOption`

File: v2/yarpcerror/errors.go [hunk 9]
```diff
@@ -25,31 +25,52 @@ import (
...
+	}
+	return status
 }
 
 // FromError returns the Status for the provided error. If the provided error
 // is not a Status, a new error with code CodeUnknown is returned.
 //
 // Returns nil if the provided error is nil.
-func FromError(err error) *Status {
+func FromError(err error) *status {
```
Comment (AlexAPB) [inline v2/yarpcerror/errors.go]: We'll need to revisit this use case / maybe return an interface here? 🤔

**Comment by kriskowal** [comment]
I would like us to expose an `Error` interface that has all the public methods of the private `*status` type. It’s reasonable for, e.g., Thrift exception objects to implement the entire `Error` interface without needing a `*status` object at all.

File: v2/yarpcerror/errors.go (hunk 9 above)
Comment (mh-park) [reply v2/yarpcerror/errors.go]: I don't quite follow. This is supposed to be a helper function to cast the error back to a status, and without this, there's no way to do so. Unless.. we expose the `Error` interface that Kris wants. But wasn't that the whole debate we had on Thursday?

**Comment by mh-park** [comment]
@kriskowal I think we discussed this on Thursday when you weren't around. We were taking a direction that doesn't use an interface, but looking at it again, I think we need this interface. Opening this up again for debate :P @AlexAPB @amckinney @robbertvanginkel

**Comment by AlexAPB** [comment]
> @kriskowal I think we discussed this on Thursday when you weren't around. We were taking a direction that doesn't use an interface, but looking at it again, I think we need this interface. Opening this up again for debate :P @AlexAPB @amckinney @robbertvanginkel

Making an interface sounds good/required to me. Since `status` is now private, we need to expose some way of inspecting what's stored in that object. Transports will only act on the methods exposed on the interface.

Let's go with the interface.

**Comment by vyshah** [comment]
Since we're basically allowing all fields on `yarpc.status` to be set at construction and exposing IsStatus, all we're really buying by making it private is immutability. I don't think this is immediately obvious but would be good to point out somewhere in the documentation.

**Comment by AlexAPB** [comment]
> Since we're basically allowing all fields on `yarpcerror.status` to be set at construction and exposing IsStatus, all we're really buying by making it private is immutability. I don't think this is immediately obvious but would be good to point out somewhere in the documentation.

We also enable others to implement the interface that isn't a `yarpcerror.status`; eg in the thrift case as kris mentioned above, the generated thrift exception doesn't have to rely on `status` at all. But yes, you're right. immutability seems desirable for errors.

Also, we should probably drop `IsStatus`, and rely on the transport layers to try casting appropriately

**Comment by mh-park** [comment]
> Also, we should probably drop `IsStatus`, and rely on the transport layers to try casting appropriately

We can't cast since `yarpcerror.status` is private, unless you mean casting to the interface.

**Comment by AlexAPB** [comment]
> > Also, we should probably drop `IsStatus`, and rely on the transport layers to try casting appropriately
> 
> We can't cast since `yarpcerror.status` is private, unless you mean casting to the interface.

exactly what I meant :)

File: v2/yarpcerror/errors.go [hunk 10]
```diff
@@ -25,293 +25,312 @@ import (
...
-// by calling Newf(CodeResourceExhausted, format, args...).
+// by calling New(CodeResourceExhausted, sprintf(format, args...)).
 func ResourceExhaustedErrorf(format string, args ...interface{}) error {
-	return Newf(CodeResourceExhausted, format, args...)
+	return New(CodeResourceExhausted, sprintf(format, args...))
 }
 
 // FailedPreconditionErrorf returns a new Status with code CodeFailedPrecondition
-// by calling Newf(CodeFailedPrecondition, format, args...).
+// by calling New(CodeFailedPrecondition, sprintf(format, args...)).
```
Comment (AlexAPB) [inline v2/yarpcerror/errors.go]: let's also drop Status from the docs here

File: v2/yarpchttp/handler.go [hunk 11]
```diff
@@ -65,33 +66,36 @@ func (h handler) ServeHTTP(w http.ResponseWriter, httpReq *http.Request) {
...
-	status := yarpcerror.FromError(yarpcerror.WrapHandlerError(err, service, procedure))
+	status := yarpcerror.WrapHandlerError(err, service, procedure)
+	errorInfo := yarpcerror.ExtractInfo(status)
 	responseWriter.SetApplicationError()
 
-	if statusCodeText, marshalErr := status.Code().MarshalText(); marshalErr != nil {
-		status = yarpcerror.Newf(yarpcerror.CodeInternal, "error %s had code %v which is unknown", status.Error(), status.Code())
+	if statusCodeText, marshalErr := errorInfo.Code.MarshalText(); marshalErr != nil {
+		status = yarpcerror.New(yarpcerror.CodeInternal, fmt.Sprintf("error %s had code %v which is unknown", status.Error(), errorInfo.Code))
+		errorInfo.Code = yarpcerror.CodeInternal
```
Comment (AlexAPB) [inline v2/yarpchttp/handler.go]: Use the `yarpcerror.Info(err)` method here and replace the 'errorInfo' value

File: v2/yarpchttp/outbound.go [hunk 12]
```diff
@@ -148,8 +147,11 @@ func (o *Outbound) call(ctx context.Context, req *yarpc.Request, reqBuf *yarpc.B
 		return nil, nil, err
 	}
 
+	var appErr error
 	if httpRes.Header.Get(ApplicationStatusHeader) == ApplicationErrorStatus {
-		res.ApplicationError = getYARPCErrorFromResponse(httpRes, resBuf, true)
+		appErr = getYARPCErrorFromResponse(httpRes, resBuf, true)
+		errorInfo := yarpcerror.ExtractInfo(appErr)
+		res.ApplicationErrorInfo = &errorInfo
```
Comment (AlexAPB) [inline v2/yarpchttp/outbound.go]: let's inline errorInfo if we don't need the variable elsewhere

File: v2/yarpctchannel/outbound.go [hunk 13]
```diff
@@ -177,16 +176,16 @@ func callWithPeer(ctx context.Context, req *yarpc.Request, reqBody *yarpc.Buffer
 			"does not match the service name received in the response: sent %q, got: %q", req.Service, resSvcName)
 	}
 
-	var appErr error
+	var appErrInfo *yarpcerror.Info
 	if res.ApplicationError() {
 		// TODO(mhp): this is a filler error for now to preserve current yarpc
 		// error behavior. This should later be more well-defined.
-		appErr = errors.New("application error")
+		appErrInfo = &yarpcerror.Info{}
```
Comment (AlexAPB) [inline v2/yarpctchannel/outbound.go]: should this have `CodeUnkown` and message set to "application error"?

File: v2/yarpcgrpc/outbound.go [hunk 14]
```diff
@@ -95,10 +94,12 @@ func (o *Outbound) Call(ctx context.Context, req *yarpc.Request, reqBuf *yarpc.B
...
 	return &yarpc.Response{
-			Peer:             metadataToPeer(responseMD),
-			Headers:          responseHeaders,
-			ApplicationError: metadataToApplicationError(invokeErr, responseMD),
+			Peer:                 metadataToPeer(responseMD),
+			Headers:              responseHeaders,
+			ApplicationErrorInfo: &errorInfo,
 		},
 		yarpc.NewBufferBytes(responseBody),
 		invokeErr
```
Comment (amckinney) [inline v2/yarpcgrpc/outbound.go]: It looks like we are losing information between the creation of the `appErr` and extracting the info adding it to the `Response`. Do we care about the information lost in `appErr`?

File: v2/yarpcgrpc/outbound.go [hunk 15]
```diff
@@ -194,7 +195,7 @@ func invokeErrorToYARPCError(err error, responseMD metadata.MD) error {
 	status, ok := status.FromError(err)
 	// if not a yarpc error or grpc error, just return a wrapped error
 	if !ok {
-		return yarpcerror.FromError(err)
+		return yarpcerror.WrapError(err)
```
Comment (kriskowal) [inline v2/yarpcgrpc/outbound.go]: As with @AlexAPB’s suggestion of removing the Error stutter from yarpcerror.Extract, I think it makes sense to yarpcerror.Wrap(err). Take or leave.

I know this makes me a terrible person, but among functional programmers, this function would probably be called yarpcerror.Lift(err) because it lifts the error type to the yarpc error level of abstraction, but does nothing if it is already at that level.

File: v2/yarpchttp/outbound.go [hunk 16]
```diff
@@ -342,10 +344,10 @@ func getYARPCErrorFromResponse(httpRes *http.Response, resBuf *yarpc.Buffer, bot
 			code = errorCode
 		}
 	}
-	return internalyarpcerror.NewWithNamef(
+	return yarpcerror.New(
 		code,
-		httpRes.Header.Get(ErrorNameHeader),
 		strings.TrimSuffix(contents, "\n"),
+		yarpcerror.WithName(httpRes.Header.Get(ErrorNameHeader)),
```
Comment (kriskowal) [inline v2/yarpchttp/outbound.go]: I’m unspeakably happy about this change.

File: v2/yarpctchannel/outbound.go (hunk 13 above)
Comment (kriskowal) [reply v2/yarpctchannel/outbound.go]: I hope `CodeUnknown` is implied by the zero object.

**Comment by kriskowal** [comment]
It’s very likely that lint is failing in CI because CI runs an old version of gofmt. I recommend downgrading your go toolchain to 0.10 for now and running make fmt to iron it out. We’ll need to modify the Docker setup to use go 0.11 going forward, but get yourself unblocked.

File: v2/yarpctchannel/outbound.go (hunk 13 above)
Comment (AlexAPB) [reply v2/yarpctchannel/outbound.go]: Unfortunately no, the default for `Code` is `CodeOk`

File: glide.yaml [hunk 17]
```diff
@@ -44,34 +41,28 @@ import:
...
-  # this is just a dependency of golang.org/x/net and if this problem is fixed
-  # then we no longer need to have this dependency in glide.yaml imports
-  # we should then move it to testImports
   repo: https://github.com/golang/tools
 - package: gopkg.in/yaml.v2
 - package: go.uber.org/multierr
   version: '>= 0.1, < 2.0'
 - package: github.com/golang/mock
   version: ^1
+- package: github.com/gogo/googleapis
```
Comment (amckinney) [inline glide.yaml]: Let's revisit this. Ideally we wouldn't need to depend on this package.

File: v2/yarpcjson/inbound_test.go [hunk 18]
```diff
@@ -149,11 +149,10 @@ func TestHandleBothResponseError(t *testing.T) {
...
-	_, resBuf, err := handleWithCodec(context.Background(), req, reqBuf, p)
-
-	require.Equal(t, errors.New("bar"), err)
+	res, resBuf, err := handleWithCodec(context.Background(), req, reqBuf, p)
 
-	var response simpleResponse
-	require.NoError(t, json.Unmarshal(resBuf.Bytes(), &response))
-	assert.Equal(t, simpleResponse{Success: true}, response)
+	require.NoError(t, err)
+	require.NotNil(t, res)
```
Comment (amckinney) [inline v2/yarpcjson/inbound_test.go]: Should we be asserting something more confidently here, as opposed to not `nil`?

File: v2/yarpcjson/outbound.go [hunk 19]
```diff
@@ -40,11 +41,11 @@ type Client struct {
...
 // Call performs an outbound JSON request.
 //
-// resBodyOut is a pointer to a value that can be filled with
-// json.Unmarshal.
+// resBodyOut and errorDetailsOut are pointers to a value that can be filled
+// with json.Unmarshal. errorDetailsOut are used for application errors.
 //
 // Returns the response or an error if the request failed.
-func (c Client) Call(ctx context.Context, procedure string, reqBody interface{}, resBodyOut interface{}, opts ...yarpc.CallOption) error {
+func (c Client) Call(ctx context.Context, procedure string, reqBody interface{}, resBodyOut interface{}, errorDetailsOut interface{}, opts ...yarpc.CallOption) error {
```
Comment (amckinney) [inline v2/yarpcjson/outbound.go]: Discussed offline - 

This is a little strange to see the `interface{}` used for both `resBodyOut` and `errorDetailsOut`, but this is necessary in JSON as a result of the new details API, i.e. users can set their details to whatever they like. If users manually configure their details to marshal into a specific type, they would pass that into the `Call` here.

File: v2/yarpcjson/outbound.go [hunk 20]
```diff
@@ -63,25 +64,41 @@ func (c Client) Call(ctx context.Context, procedure string, reqBody interface{},
...
-	var decodeErr error
 	if _, err = call.ReadFromResponse(ctx, res); err != nil {
-		decodeErr = err
+		return err
 	}
-	if resBuf != nil {
-		if err := json.NewDecoder(resBuf).Decode(resBodyOut); err != nil && decodeErr == nil {
-			decodeErr = yarpcencoding.ResponseBodyDecodeError(&req, err)
+
+	if res.ApplicationErrorInfo != nil {
```
Comment (amckinney) [inline v2/yarpcjson/outbound.go]: nit: Since we're accessing `res.ApplicationErrorInfo` a lot in this branch, we can initialize a separate, more readable variable for it:
```
if info := res.ApplicationErrorInfo; info != nil { ... }
```

File: v2/yarpchttp/outbound.go [hunk 21]
```diff
@@ -344,11 +331,12 @@ func getYARPCErrorFromResponse(httpRes *http.Response, resBuf *yarpc.Buffer, bot
...
-		strings.TrimSuffix(contents, "\n"),
-		yarpcerror.WithName(httpRes.Header.Get(ErrorNameHeader)),
-	)
+	message := httpRes.Header.Get(ErrorMessageHeader)
+	return yarpcerror.Info{
+		Code:    code,
+		Message: strings.TrimSuffix(message, "\n"),
+		Name:    httpRes.Header.Get(ErrorNameHeader),
+	}
 }
```
Comment (amckinney) [inline v2/yarpchttp/outbound.go]: I realize that this was a decision we accepted, but we should note that this is technically wire-breaking, i.e. we are always returning the error along the header as opposed to sometimes along the response `yarpc.Buffer`.

File: v2/internal/internalyarpcobservability/call.go [hunk 22]
```diff
@@ -115,15 +115,7 @@ func (c call) endStats(elapsed time.Duration, err error, isApplicationError bool
 		return
 	}
 
-	if !yarpcerror.IsStatus(err) {
-		c.edge.serverErrLatencies.Observe(elapsed)
-		if counter, err := c.edge.serverFailures.Get(_error, "unknown_internal_yarpc"); err == nil {
```
Comment (robbertvanginkel) [inline v2/internal/internalyarpcobservability/call.go]: Do we have a changelog for expected behaviour changes yarpcv1 to v2? Would be good to attach that `unknown_internal_yarpc` metrics now change to `unknown`

File: v2/yarpcgrpc/outbound.go [hunk 23]
```diff
@@ -147,8 +134,9 @@ func (o *Outbound) invoke(
 		callOptions...,
 	)
 
+	errInfo, appErrBody, err := invokeErrorToYARPCError(err, responseMD)
```
Comment (robbertvanginkel) [inline v2/yarpcgrpc/outbound.go]: Nit: maybe move this inline to 129, and have `invokeErrorToYARPCError(peer.clientConn.Invoke(...), ...)`. Makes it a little easier to see we don't forget the errorcheck after line 128.

File: v2/yarpcgrpc/outbound.go [hunk 24]
```diff
@@ -174,50 +177,48 @@ func metadataToPeer(responseMD metadata.MD) yarpc.Identifier {
...
 		message = strings.TrimPrefix(message, name+": ")
 	} else if name != "" && message == name {
 		message = ""
 	}
-	return yarpcerror.New(code, message, yarpcerror.WithName(name))
+
+	if isAppErr {
+		buf, err := proto.Marshal(s.Proto())
+		if err != nil {
+			return nil, nil, err
```
Comment (robbertvanginkel) [inline v2/yarpcgrpc/outbound.go]: Should this last error be a yarpcerror?

File: glide.yaml (hunk 17 above)
Comment (amckinney) [reply glide.yaml]: From https://jbrandhorst.com/post/grpc-errors/, this appears to be non-actionable for now. I'll investigate our dependency on `gogo/protobuf` in a separate change.

**Comment by AlexAPB** [comment]
you're a rockstar 🥇 💯 🤘 🎸

File: v2/internal/internalyarpcobservability/middleware.go [hunk 25]
```diff
@@ -64,7 +64,7 @@ func (m *Middleware) Handle(ctx context.Context, req *yarpc.Request, reqBuf *yar
 
 	isApplicationError := false
 	if res != nil {
-		isApplicationError = res.ApplicationErrorInfo != nil
+		isApplicationError = res.ErrorInfo != nil
 	}
 	// TODO(mhp): Now that we are including the application error into the
```
[the remark this answers is not in the document]
Comment (mh-park) [reply v2/internal/internalyarpcobservability/middleware.go]: Not yet, we're just saying there was an application error, and not logging any information about it. I'll do that in a follow-up.

File: v2/yarpcgrpc/outbound.go (hunk 24 above)
Comment (mh-park) [reply v2/yarpcgrpc/outbound.go]: yarpcerrors are inspected using our GetInfo or GetExtract methods only, so calling a GetInfo on a non-yarpcerror will wrap it accordingly. In other words, we don't have to actually wrap any errors with a yarpcerror, unless we want to explicitly attach an error code or name to it.

File: v2/yarpcgrpc/stream.go [hunk 26]
```diff
@@ -137,14 +137,7 @@ func toYARPCStreamError(err error) error {
...
 	}
-	if yarpcerror.IsStatus(err) {
-		return err
-	}
-	status, ok := status.FromError(err)
-	// if not a yarpc error or grpc error, just return a wrapped error
-	if !ok {
-		return yarpcerror.WrapError(err)
-	}
+	status, _ := status.FromError(err)
```
[the remark this answers is not in the document]
Comment (mh-park) [reply v2/yarpcgrpc/stream.go]: Previously, it was a "if not ok, set code to unknown". The code below this catches this case.

File: v2/yarpcgrpc/outbound.go [hunk 27]
```diff
@@ -174,40 +177,48 @@ func metadataToPeer(responseMD metadata.MD) yarpc.Identifier {
...
 	if responseMD != nil {
 		value, ok := responseMD[ErrorNameHeader]
 		// TODO: what to do if the length is > 1?
 		if ok && len(value) == 1 {
 			name = value[0]
 		}
+
+		appErrHeader, ok := responseMD[ApplicationErrorHeader]
+		if ok && len(appErrHeader) == 1 {
+			isAppErr = appErrHeader[0] == ApplicationErrorHeaderValue
```
[the remark this answers is not in the document]
Comment (mh-park) [reply v2/yarpcgrpc/outbound.go]: This means I can do:
```
_, isAppErr = responseMD[ApplicationErrorHeader]
```
instead, correct?

File: v2/yarpctchannel/handler.go [hunk 28]
```diff
@@ -161,26 +161,34 @@ func (h handler) handleKernel(ctx context.Context, start time.Time, req *yarpc.R
...
-	if retErr != nil && res != nil && res.ApplicationErrorInfo != nil {
-		// We have an error, so we're going to propagate it as a yarpc error,
-		// regardless of whether or not it is a system error.
+	// This denotes that there was an application error.
+	if res.ApplicationErrorInfo != nil {
 		errorInfo := res.ApplicationErrorInfo
-		text, err := errorInfo.Code.MarshalText()
+		code := errorInfo.Code
+		// Black-hole requests on resource exhausted errors.
+		if code == yarpcerror.CodeResourceExhausted {
```
[the remark this answers is not in the document]
Comment (mh-park) [reply v2/yarpctchannel/handler.go]: :/
